crate2nix/Cargo.nix

```diff
@@ -2368,13 +2358,10 @@ rec {
               target = target // { test = runTests; };
             }
           );
-        buildByPackageId = packageId: buildByPackageIdImpl packageId;
-
-        # Memoize built packages so that reappearing packages are only built once.
```

Comment:
Maybe we can do this again now that the hash thing has been changed in https://github.com/NixOS/nixpkgs/pull/105305?

On the other hand, isn't the comment wrong? I would assume this is just saving eval time, and the derivations will be identical and thus deduplicated either way.

Reply:
My understanding was also that this was an eval optimization.

I don't think that PR affects this. `packageId` doesn't have anything to do with the hash; IIRC `packageId` is just the name and possibly the version if there are multiple. Even if you were to extend it to include the platform, it is not obvious which packages need to be built for which platforms until you walk through the dependency tree.

Reply:
OK fair point, then I think it might be nice to do the same corecursive stuff we do in nixpkgs?
```nix
let
buildByPackageIdForPkgsMemo = mkBuildByPackageIdForPkgsMemo pkgs;
mkBuildByPackageIdForPkgsMemo = pkgs: {
  cratePkgs = lib.mapAttrs (packageId: value: buildByPackageIdForPkgs pkgs packageId) crateConfigs;
  build = mkBuildByPackageIdForPkgsMemo pkgs.buildPackages;
};
```
It's fine that this contains too much stuff, again cause lazy eval. You use `cratePkgs` for dependencies, `build.cratePkgs` for dev-dependencies, and a nasty backtracking of switching from `cratePkgs` to `build.cratePkgs` for any proc macro crate. (Shame on them for not making proc macros a different type of dependency! I did mention it when it came out...)